myia/anf_ir.py

```diff
@@ -0,0 +1,304 @@
+        """Get an input by its index."""
+        return self.data[index]
+
+    @overload
+    def __setitem__(self, index: int, value: ANFNode) -> None:
+        pass
+
+    @overload  # noqa: F811
+    def __setitem__(self, index: slice, value: Iterable[ANFNode]) -> None:
+        pass
+
+    def __setitem__(self, index, value):  # noqa: F811
+        """Replace an input with another."""
+        if isinstance(index, slice):
+            raise ValueError("slice assignment not supported")
+        if index < 0:
+            index += len(self)
+        old_value = self.data[index]
+        old_value.uses.remove((self.node, index))
+        value.uses.add((self.node, index))
+        self.data[index] = value
+
+    @overload
+    def __delitem__(self, index: int) -> None:
+        pass
+
+    @overload  # noqa: F811
+    def __delitem__(self, index: slice) -> None:
+        pass
+
+    def __delitem__(self, index):  # noqa: F811
+        """Delete an input."""
+        if isinstance(index, slice):
+            raise ValueError("slice deletion not supported")
+        if index < 0:
+            index += len(self)
+        value = self.data[index]
+        value.uses.remove((self.node, index))
+        for i, next_value in enumerate(self.data[index + 1:]):
+            next_value.uses.remove((self.node, i + index + 1))
+            next_value.uses.add((self.node, i + index))
+        del self.data[index]
+
+    def __len__(self) -> int:
+        """Get the number of inputs."""
+        return len(self.data)
+
+    def insert(self, index: int, value: ANFNode) -> None:
+        """Insert an input at a given location."""
+        if index < 0:
+            index += len(self)
+        for i, next_value in enumerate(reversed(self.data[index:])):
+            next_value.uses.remove((self.node, len(self) - i - 1))
+            next_value.uses.add((self.node, len(self) - i))
+        value.uses.add((self.node, index))
+        self.data.insert(index, value)
+
+    def __repr__(self) -> str:
+        """Return a string representation of the inputs."""
+        return f"Inputs({self.data})"
+
+    def __eq__(self, other) -> bool:
+        """Test whether a list of inputs is equal to another list.
+
+        Note:
+            The goal of `Inputs` is to behave exactly like a list, but track
+            insertions and deletions to keep the bidirectional graph structure
+            up to date. As such, which node an `Inputs` object is attached to
+            is irrelevant to the equality test. Instead, a simple element by
+            element test is performed.
+
+        """
+        return all(x == y for x, y in zip(self, other))
+
+
+class Apply(ANFNode):
+    """A function application.
+
+    This node represents the application of a function to a set of arguments.
+
+    """
+
+    def __init__(self, inputs: List[ANFNode], graph: 'Graph') -> None:
+        """Construct an application."""
+        super().__init__(inputs, None, graph)
```

Comment:
The `value` given to `__init__` should not be `None`.

Reply:
What should it be then? So far the convention has been `node.value == None` for `Apply`, `node.value == PARAMETER` for `Parameter`, and `node.value` is whatever for `Constant`. Want me to replace it with a special `APPLY` value?

Reply:
Some special value, yes, because `None` *could* be a valid value for a constant.

Reply:
Done.